File: 2023/day 01/day_01.py
"""
Advent of Code 2023 - Day 01
Pieter Kitslaar
"""

from pathlib import Path
THIS_DIR = Path(__file__).parent

def data():
    with open(THIS_DIR / 'input.txt') as f:
        return f.read()

EXAMPLE_DATA = """\
1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
"""

import re
# find only digits
DIGIT_RE = re.compile('\d')   

# find digits (e.g. 1,2,3 ) or digit words (e.g. one, two, three)
# The value for normal digits is None so we can test on it later
WORD_VALUES = {'\d': None,
                'one': 1, 'two': 2, 'three': 3, 'four': 4, 'five': 5, 
               'six': 6, 'seven': 7, 'eight': 8, 'nine': 9, 'zero': 0}
DIGIT_WORD_RE = re.compile("|".join(f"{dw}" for dw in WORD_VALUES.keys()))
def solve(data, part2=False):
    result = 0
    RE = DIGIT_WORD_RE if part2 else DIGIT_RE
    for l in data.splitlines():
        # start of search
        s = 0
        digit_words = []
        while m := RE.search(l, s):
            digit_words.append(m.group())
            s = m.span()[1]
            # we want overlapping search so for longer matches we 
            # set the search start back one character
            if len(digit_words[-1]) > 1: 
                s -= 1
        digits = [str(WORD_VALUES.get(dw, dw)) for dw in digit_words]
        result += int(digits[0] + digits[-1])
    return result

def test_example():
    result = solve(EXAMPLE_DATA)
    print(f'example: {result}')
    assert result == 142

def test_part1():
    result = solve(data())
    print('Part 1:', result)
    assert result == 55017

EXAMPLE_DATA2 = """\
two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen
"""	

def test_example2():
    result = solve(EXAMPLE_DATA2, part2=True)
    print(f'example 2: {result}')
    assert result == 281

def test_part2():
    result = solve(data(), part2=True)
    print('Part 2:', result)
    assert result == 53539

File: 2023/day 01/day_01.rs
const EXAMPLE_DATA: &str = indoc::indoc! {"
1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
"};

const EXAMPLE_DATA2: &str = indoc::indoc! {"
two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen
"};

extern crate regex;
use regex::Regex;

fn solve(data: &str, part2: bool) -> Option<String> {
    let mut sum = 0;

    let mut word_values = std::collections::HashMap::new();
    word_values.insert(r"\d", "");
    if part2 {
        word_values.insert("one", "1");
        word_values.insert("two", "2");
        word_values.insert("three", "3");
        word_values.insert("four", "4");
        word_values.insert("five", "5");
        word_values.insert("six", "6");
        word_values.insert("seven", "7");
        word_values.insert("eight", "8");
        word_values.insert("nine", "9");
        word_values.insert("zero", "0");
    }

    // Combine all keys in WORD_VALUES into a single regex
    let word_re = Regex::new(
        word_values
            .keys()
            .cloned()
            .collect::<Vec<&str>>()
            .join("|")
            .as_str(),
    )
    .unwrap();

    for line in data.lines() {
        let mut digits = Vec::new();
        let mut start = 0;
        loop {
            if let Some(mat) = word_re.find(&line[start..]) {
                let match_str = mat.as_str();
                let match_value = word_values.get(match_str).unwrap_or(&match_str);
                digits.push(*match_value);
                start += mat.start() + match_str.len();
                if match_str.len() > 1 {
                    start -= 1
                }
            } else {
                break;
            }
        }
        let combined = format!("{}{}", digits.first()?, digits.last()?);
        sum += combined.parse::<u32>().ok()?;
    }
    Some(sum.to_string())
}

fn example() {
    let result = solve(EXAMPLE_DATA, false).unwrap();
    println!("example: {}", result);
    assert_eq!(result, "142");
}

#[test]
fn test_example() {
    example()
}

fn part1() {
    let data = data();
    let result = solve(data, false).unwrap();
    println!("Part 1: {}", result);
    assert_eq!(result, "55017");
}

#[test]
fn test_part1() {
    part1()
}

fn example2() {
    let result = solve(EXAMPLE_DATA2, true).unwrap();
    println!("example2: {}", result);
    assert_eq!(result, "281");
}

#[test]
fn test_example2() {
    example2()
}

fn part2() {
    let data = data();
    let result = solve(data, true).unwrap();
    println!("Part 2: {}", result);
    assert_eq!(result, "53539");
}

#[test]
fn test_part2() {
    part2()
}

fn data() -> &'static str {
    let data = include_str!("./input.txt");
    data
}

fn main() {
    example();
    part1();
    example2();
    part2();
}
